File: ScapesCore/src/main/java/org/tobi29/scapes/chunk/terrain/infinite/TerrainInfiniteRendererChunk.java
/*
 * Copyright 2012-2016 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.tobi29.scapes.chunk.terrain.infinite;

import org.tobi29.scapes.engine.graphics.*;
import org.tobi29.scapes.engine.utils.ArrayUtil;
import org.tobi29.scapes.engine.utils.Pair;
import org.tobi29.scapes.engine.utils.Streams;
import org.tobi29.scapes.engine.utils.graphics.Cam;
import org.tobi29.scapes.engine.utils.math.AABB;
import org.tobi29.scapes.engine.utils.math.FastMath;

import java.util.Arrays;
import java.util.concurrent.atomic.AtomicBoolean;

public class TerrainInfiniteRendererChunk {
    private final TerrainInfiniteChunkClient chunk;
    private final TerrainInfiniteRenderer renderer;
    private final TerrainInfiniteChunkModel[] vao;
    private final AtomicBoolean[] geometryDirty;
    private final boolean[] geometryInit, solid, visible, prepareVisible,
            culled, lod;

    public TerrainInfiniteRendererChunk(TerrainInfiniteChunkClient chunk,
            TerrainInfiniteRenderer renderer) {
        this.chunk = chunk;
        this.renderer = renderer;
        int zSections = chunk.zSize() >> 4;
        geometryDirty = new AtomicBoolean[zSections];
        ArrayUtil.fill(geometryDirty, () -> new AtomicBoolean(true));
        geometryInit = new boolean[zSections + 1];
        vao = new TerrainInfiniteChunkModel[zSections];
        lod = new boolean[zSections];
        solid = new boolean[zSections];
        visible = new boolean[zSections];
        prepareVisible = new boolean[zSections];
        culled = new boolean[zSections];
        Arrays.fill(solid, true);
        renderer.addToQueue(this);
    }

    public TerrainInfiniteChunkClient chunk() {
        return chunk;
    }

    public int zSections() {
        return vao.length;
    }

    public void render(GL gl, Shader shader1, Shader shader2, Cam cam) {
        double relativeX = chunk.blockX() - cam.position.doubleX();
        double relativeY = chunk.blockY() - cam.position.doubleY();
        MatrixStack matrixStack = gl.matrixStack();
        for (int i = 0; i < vao.length; i++) {
            double relativeZ = (i << 4) - cam.position.doubleZ();
            double distance =
                    FastMath.sqr(relativeX + 8) + FastMath.sqr(relativeY + 8) +
                            FastMath.sqr(relativeZ + 8);
            boolean newLod = distance < 9216;
            TerrainInfiniteChunkModel vao = this.vao[i];
            if (vao != null && vao.model.isPresent()) {
                Pair<Model, AABB> m = vao.model.get();
                if (vao.lod != newLod) {
                    setGeometryDirty(i);
                }
                if (cam.frustum.inView(m.b) != 0) {
                    boolean animated = distance < 2304;
                    Matrix matrix = matrixStack.push();
                    matrix.translate((float) relativeX, (float) relativeY,
                            (float) relativeZ);
                    if (!m.a.render(gl, animated ? shader1 : shader2)) {
                        setGeometryDirty(i);
                    }
                    matrixStack.pop();
                } else {
                    m.a.ensureStored(gl);
                }
            }
        }
    }

    public void renderAlpha(GL gl, Shader shader1, Shader shader2, Cam cam) {
        double relativeX = chunk.blockX() - cam.position.doubleX();
        double relativeY = chunk.blockY() - cam.position.doubleY();
        MatrixStack matrixStack = gl.matrixStack();
        for (int i = 0; i < vao.length; i++) {
            TerrainInfiniteChunkModel vao = this.vao[i];
            if (vao != null && vao.modelAlpha.isPresent()) {
                Pair<Model, AABB> m = vao.modelAlpha.get();
                if (cam.frustum.inView(m.b) != 0) {
                    double relativeZ = (i << 4) - cam.position.doubleZ();
                    double distance = FastMath.sqr(relativeX + 8) +
                            FastMath.sqr(relativeY + 8) +
                            FastMath.sqr(relativeZ + 8);
                    boolean animated = distance < 2304;
                    Matrix matrix = matrixStack.push();
                    matrix.translate((float) relativeX, (float) relativeY,
                            (float) relativeZ);
                    if (!m.a.render(gl, animated ? shader1 : shader2)) {
                        setGeometryDirty(i);
                    }
                    matrixStack.pop();
                } else {
                    m.a.ensureStored(gl);
                }
            }
        }
    }

    public boolean isLoaded() {
        return geometryInit[0] && chunk.isLoaded();
    }

    public void renderFrame(GL gl, Model frame, Shader shader, Cam cam) {
        MatrixStack matrixStack = gl.matrixStack();
        for (int i = 0; i < vao.length; i++) {
            TerrainInfiniteChunkModel vao = this.vao[i];
            if (vao != null && vao.model.isPresent()) {
                AABB aabb = vao.model.get().b;
                Matrix matrix = matrixStack.push();
                gl.setAttribute2f(4, 1.0f, 1.0f);
                if (!chunk.isLoaded()) {
                    gl.setAttribute4f(GL.COLOR_ATTRIBUTE, 1.0f, 0.0f, 0.0f,
                            1.0f);
                } else if (geometryDirty[i].get()) {
                    gl.setAttribute4f(GL.COLOR_ATTRIBUTE, 1.0f, 1.0f, 0.0f,
                            1.0f);
                } else {
                    gl.setAttribute4f(GL.COLOR_ATTRIBUTE, 0.0f, 1.0f, 0.0f,
                            1.0f);
                }
                matrix.translate((float) (aabb.minX - cam.position.doubleX()),
                        (float) (aabb.minY - cam.position.doubleY()),
                        (float) (aabb.minZ - cam.position.doubleZ()));
                matrix.scale((float) (aabb.maxX - aabb.minX),
                        (float) (aabb.maxY - aabb.minY),
                        (float) (aabb.maxZ - aabb.minZ));
                frame.render(gl, shader);
                matrixStack.pop();
            }
        }
    }

    public synchronized void replaceMesh(int i,
            TerrainInfiniteChunkModel model) {
        if (visible[i] && model != null) {
            model.model.ifPresent(m -> m.a.setWeak(true));
            model.modelAlpha.ifPresent(m -> m.a.setWeak(true));
            model.model.ifPresent(
                    m -> m.b.add(chunk.blockX(), chunk.blockY(), i << 4));
            model.modelAlpha.ifPresent(
                    m -> m.b.add(chunk.blockX(), chunk.blockY(), i << 4));
            vao[i] = model;
        } else {
            vao[i] = null;
        }
        if (!geometryInit[0]) {
            geometryInit[i + 1] = true;
            boolean flag = true;
            for (int j = 1; j < geometryInit.length; j++) {
                if (!geometryInit[j]) {
                    flag = false;
                    break;
                }
            }
            if (flag) {
                geometryInit[0] = true;
            }
        }
    }

    public void setGeometryDirty() {
        for (int i = 0; i < vao.length; i++) {
            geometryDirty[i].set(true);
        }
        renderer.addToQueue(this);
    }

    public void setGeometryDirty(int i) {
        if (!geometryDirty[i].getAndSet(true)) {
            renderer.addToQueue(this, i);
        }
    }

    public void setSolid(int i, boolean value) {
        if (i >= 0 && i < solid.length) {
            solid[i] = value;
        }
    }

    protected void resetPrepareVisible() {
        Arrays.fill(prepareVisible, false);
    }

    protected void setPrepareVisible(int i) {
        if (i >= 0 && i < prepareVisible.length) {
            prepareVisible[i] = true;
        }
    }

    public void updateVisible() {
        for (int i = 0; i < visible.length; i++) {
            boolean oldVisible = visible[i];
            if (prepareVisible[i] && !oldVisible) {
                visible[i] = true;
                setGeometryDirty(i);
            } else if (!prepareVisible[i] && oldVisible) {
                visible[i] = false;
                vao[i] = null;
            }
        }
    }

    public void setCulled(boolean value) {
        Arrays.fill(culled, value);
    }

    public boolean setCulled(int i, boolean value) {
        if (i >= 0 && i < culled.length) {
            boolean old = culled[i];
            culled[i] = value;
            return old != value;
        }
        return false;
    }

    public boolean unsetGeometryDirty(int i) {
        return geometryDirty[i].getAndSet(false);
    }

    public boolean isSolid(int i) {
        return i < 0 || i >= solid.length || solid[i];
    }

    public boolean isVisible(int i) {
        return !(i < 0 || i >= visible.length) && visible[i];
    }

    public boolean isCulled(int i) {
        return !(i < 0 || i >= culled.length) && culled[i];
    }

    public void reset() {
        Arrays.fill(vao, null);
        Arrays.fill(solid, true);
        Streams.forEach(geometryDirty, value -> value.set(true));
        Arrays.fill(geometryInit, false);
        Arrays.fill(culled, false);
        renderer.addToQueue(this);
    }
}

File: ScapesCore/src/main/kotlin/org/tobi29/scapes/chunk/terrain/infinite/TerrainInfiniteRendererChunk.kt
/*
 * Copyright 2012-2017 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.chunk.terrain.infinite

import org.tobi29.arrays.fill
import org.tobi29.graphics.Cam
import org.tobi29.scapes.engine.graphics.GL
import org.tobi29.scapes.engine.graphics.Model
import org.tobi29.scapes.engine.graphics.Shader
import org.tobi29.scapes.engine.graphics.push
import org.tobi29.stdex.atomic.AtomicBoolean
import org.tobi29.stdex.math.sqr

class TerrainInfiniteRendererChunk(private val chunk: TerrainInfiniteChunkClient,
                                   private val renderer: TerrainInfiniteRenderer) {
    private val vao: Array<TerrainInfiniteChunkModel?>
    private val geometryDirty: Array<AtomicBoolean>
    private val geometryInit: BooleanArray
    private val solid: BooleanArray
    private val visible: BooleanArray
    private val prepareVisible: BooleanArray
    private val culled: BooleanArray

    init {
        val zSections = chunk.size.z shr 4
        geometryDirty = Array(zSections) { AtomicBoolean(true) }
        geometryInit = BooleanArray(zSections + 1)
        vao = arrayOfNulls(zSections)
        solid = BooleanArray(zSections)
        visible = BooleanArray(zSections)
        prepareVisible = BooleanArray(zSections)
        culled = BooleanArray(zSections) { true }
        renderer.addToQueue(this)
    }

    fun chunk(): TerrainInfiniteChunkClient {
        return chunk
    }

    fun zSections(): Int {
        return vao.size
    }

    fun render(gl: GL,
               shader1: Shader,
               shader2: Shader,
               cam: Cam) {
        val relativeX = chunk.posBlock.x - cam.position.x
        val relativeY = chunk.posBlock.y - cam.position.y
        for (i in vao.indices) {
            val relativeZ = (i shl 4) - cam.position.z
            val distance = sqr(relativeX + 8) +
                    sqr(relativeY + 8) +
                    sqr(relativeZ + 8)
            val newLod = distance < sqr(renderer.lodDistance)
            val vao = this.vao[i]
            if (vao != null && vao.model != null) {
                if (vao.lod != newLod) {
                    setGeometryDirty(i)
                }
                if (cam.frustum.inView(vao.model.second) != 0) {
                    val animated = distance < 2304
                    gl.matrixStack.push { matrix ->
                        matrix.translate(relativeX.toFloat(),
                                relativeY.toFloat(), relativeZ.toFloat())
                        if (!vao.model.first.render(gl,
                                if (animated) shader1 else shader2)) {
                            setGeometryDirty(i)
                        }
                    }
                } else {
                    vao.model.first.ensureStored(gl)
                }
            }
        }
    }

    fun renderAlpha(gl: GL,
                    shader1: Shader,
                    shader2: Shader,
                    cam: Cam) {
        val relativeX = chunk.posBlock.x - cam.position.x
        val relativeY = chunk.posBlock.y - cam.position.y
        for (i in vao.indices) {
            val vao = this.vao[i]
            if (vao != null && vao.modelAlpha != null) {
                if (cam.frustum.inView(vao.modelAlpha.second) != 0) {
                    val relativeZ = (i shl 4) - cam.position.z
                    val distance = sqr(relativeX + 8) +
                            sqr(relativeY + 8) +
                            sqr(relativeZ + 8)
                    val animated = distance < 2304
                    gl.matrixStack.push { matrix ->
                        matrix.translate(relativeX.toFloat(),
                                relativeY.toFloat(), relativeZ.toFloat())
                        if (!vao.modelAlpha.first.render(gl,
                                if (animated) shader1 else shader2)) {
                            setGeometryDirty(i)
                        }
                    }
                } else {
                    vao.modelAlpha.first.ensureStored(gl)
                }
            }
        }
    }

    val isLoaded: Boolean
        get() = geometryInit[0] && chunk.isLoaded

    fun renderFrame(gl: GL,
                    frame: Model,
                    shader: Shader,
                    cam: Cam) {
        for (i in vao.indices) {
            val vao = this.vao[i]
            if (vao != null && vao.model != null) {
                gl.matrixStack.push { matrix ->
                    gl.setAttribute2f(4, 1.0f, 1.0f)
                    if (!chunk.isLoaded) {
                        gl.setAttribute4f(GL.COLOR_ATTRIBUTE, 1.0f, 0.0f, 0.0f,
                                1.0f)
                    } else if (geometryDirty[i].get()) {
                        gl.setAttribute4f(GL.COLOR_ATTRIBUTE, 1.0f, 1.0f, 0.0f,
                                1.0f)
                    } else {
                        gl.setAttribute4f(GL.COLOR_ATTRIBUTE, 0.0f, 1.0f, 0.0f,
                                1.0f)
                    }
                    matrix.translate(
                            (vao.model.second.minX - cam.position.x).toFloat(),
                            (vao.model.second.minY - cam.position.y).toFloat(),
                            (vao.model.second.minZ - cam.position.z).toFloat())
                    matrix.scale(
                            (vao.model.second.maxX - vao.model.second.minX).toFloat(),
                            (vao.model.second.maxY - vao.model.second.minY).toFloat(),
                            (vao.model.second.maxZ - vao.model.second.minZ).toFloat())
                    frame.render(gl, shader)
                }
            }
        }
    }

    @Synchronized
    fun replaceMesh(i: Int,
                    model: TerrainInfiniteChunkModel?) {
        if (visible[i] && model != null) {
            model.model?.let { it.first.weak = true }
            model.modelAlpha?.let { it.first.weak = true }
            model.model?.second?.add(chunk.posBlock.x.toDouble(),
                    chunk.posBlock.y.toDouble(),
                    (i shl 4).toDouble())
            model.modelAlpha?.second?.add(chunk.posBlock.x.toDouble(),
                    chunk.posBlock.y.toDouble(),
                    (i shl 4).toDouble())
            vao[i] = model
        } else {
            vao[i] = null
        }
        if (!geometryInit[0]) {
            geometryInit[i + 1] = true
            if (geometryInit.indices.any { geometryInit[it] }) {
                geometryInit[0] = true
            }
        }
    }

    fun setGeometryDirty() {
        for (i in vao.indices) {
            geometryDirty[i].set(true)
        }
        renderer.addToQueue(this)
    }

    fun setGeometryDirty(i: Int) {
        if (!geometryDirty[i].getAndSet(true)) {
            renderer.addToQueue(this, i)
        }
    }

    fun setSolid(i: Int,
                 value: Boolean) {
        if (i >= 0 && i < solid.size) {
            solid[i] = value
        }
    }

    fun resetPrepareVisible() {
        prepareVisible.fill { false }
    }

    fun setPrepareVisible(i: Int) {
        if (i >= 0 && i < prepareVisible.size) {
            prepareVisible[i] = true
        }
    }

    fun updateVisible() {
        for (i in visible.indices) {
            val oldVisible = visible[i]
            if (prepareVisible[i] && !oldVisible) {
                visible[i] = true
                setGeometryDirty(i)
            } else if (!prepareVisible[i] && oldVisible) {
                visible[i] = false
                vao[i] = null
            }
        }
    }

    fun setCulled(value: Boolean) {
        culled.fill { value }
    }

    fun setCulled(i: Int,
                  value: Boolean): Boolean {
        if (i >= 0 && i < culled.size) {
            val old = culled[i]
            culled[i] = value
            return old != value
        }
        return false
    }

    fun unsetGeometryDirty(i: Int): Boolean {
        return geometryDirty[i].getAndSet(false)
    }

    fun isSolid(i: Int): Boolean {
        return i < 0 || i >= solid.size || solid[i]
    }

    fun isVisible(i: Int): Boolean {
        return !(i < 0 || i >= visible.size) && visible[i]
    }

    fun isCulled(i: Int): Boolean {
        return !(i < 0 || i >= culled.size) && culled[i]
    }

    fun reset() {
        vao.fill { null }
        solid.fill { true }
        geometryDirty.forEach { it.set(true) }
        geometryInit.fill { false }
        culled.fill { false }
        renderer.addToQueue(this)
    }
}
